File: app/src/main/java/io/github/tobyhs/weatherweight/data/model/ForecastResultSet.java
package io.github.tobyhs.weatherweight.data.model;

import java.util.List;

import com.google.auto.value.AutoValue;
import me.tatarka.gsonvalue.annotations.GsonBuilder;
import org.threeten.bp.ZonedDateTime;

/**
 * A result set for weather forecasts
 */
@AutoValue
public abstract class ForecastResultSet {
    /**
     * @return location this forecast is for
     */
    public abstract String getLocation();

    /**
     * @return publication time of this forecast
     */
    public abstract ZonedDateTime getPublicationTime();

    /**
     * @return daily forecasts for today and upcoming days
     */
    public abstract List<DailyForecast> getForecasts();

    @GsonBuilder
    public static Builder builder() {
        return new AutoValue_ForecastResultSet.Builder();
    }

    @AutoValue.Builder
    public abstract static class Builder {
        public abstract Builder setLocation(String location);
        public abstract Builder setPublicationTime(ZonedDateTime publicationTime);
        public abstract Builder setForecasts(List<DailyForecast> forecasts);
        public abstract ForecastResultSet build();
    }
}

File: app/src/main/java/io/github/tobyhs/weatherweight/data/model/ForecastResultSet.kt
package io.github.tobyhs.weatherweight.data.model

import com.squareup.moshi.JsonClass

import java.time.ZonedDateTime

/**
 * A result set for weather forecasts
 *
 * @property location location this forecast is for
 * @property publicationTime time of this forecast
 * @property forecasts daily forecasts for today and upcoming days
 */
@JsonClass(generateAdapter = true)
data class ForecastResultSet(
    val location: String,
    val publicationTime: ZonedDateTime,
    val forecasts: List<DailyForecast>,
)
